src/libs/actions/OnyxUpdates.js

```diff
@@ -1,22 +1,138 @@
 import Onyx from 'react-native-onyx';
+import _ from 'underscore';
+import PusherUtils from '../PusherUtils';
 import ONYXKEYS from '../../ONYXKEYS';
+import * as QueuedOnyxUpdates from './QueuedOnyxUpdates';
+import CONST from '../../CONST';
+
+// This key needs to be separate from ONYXKEYS.ONYX_UPDATES_FROM_SERVER so that it can be updated without triggering the callback when the server IDs are updated. If that
+// callback were triggered it would lead to duplicate processing of server updates.
+let lastUpdateIDAppliedToClient = 0;
+Onyx.connect({
+    key: ONYXKEYS.ONYX_UPDATES_LAST_UPDATE_ID_APPLIED_TO_CLIENT,
+    callback: (val) => (lastUpdateIDAppliedToClient = val),
+});
 
 /**
- *
- * @param {Number} [lastUpdateID]
- * @param {Number} [previousUpdateID]
+ * @param {Object} data
+ * @param {Object} data.request
+ * @param {Object} data.responseData
+ * @returns {Promise}
+ */
+function applyHTTPSOnyxUpdates({request, responseData}) {
+    console.debug('[OnyxUpdateManager] Applying https update');
+    // For most requests we can immediately update Onyx. For write requests we queue the updates and apply them after the sequential queue has flushed to prevent a replay effect in
+    // the UI. See https://github.com/Expensify/App/issues/12775 for more info.
+    const updateHandler = request.data.apiRequestType === CONST.API_REQUEST_TYPE.WRITE ? QueuedOnyxUpdates.queueOnyxUpdates : Onyx.update;
+
+    // First apply any onyx data updates that are being sent back from the API. We wait for this to complete and then
+    // apply successData or failureData. This ensures that we do not update any pending, loading, or other UI states contained
+    // in successData/failureData until after the component has received and API data.
+    const onyxDataUpdatePromise = responseData.onyxData ? updateHandler(responseData.onyxData) : Promise.resolve();
+
+    return onyxDataUpdatePromise
+        .then(() => {
+            // Handle the request's success/failure data (client-side data)
+            if (responseData.jsonCode === 200 && request.successData) {
+                return updateHandler(request.successData);
+            }
+            if (responseData.jsonCode !== 200 && request.failureData) {
+                return updateHandler(request.failureData);
+            }
+            return Promise.resolve();
+        })
+        .then(() => {
+            console.debug('[OnyxUpdateManager] Done applying HTTPS update');
+            return responseData;
+        });
+}
+
+/**
+ * @param {Object} data
+ * @param {Object} data.updates
+ * @returns {Promise}
+ */
+function applyPusherOnyxUpdates({updates}) {
+    console.debug('[OnyxUpdateManager] Applying pusher update');
+    const pusherEventPromises = _.reduce(
+        updates,
+        (result, update) => {
+            result.push(PusherUtils.triggerMultiEventHandler(update.eventType, update.data));
+            return result;
+        },
+        [],
+    );
```

Comment:
```suggestion
    const pusherEventPromises = _.map(updates, update => PusherUtils.triggerMultiEventHandler(update.eventType, update.data));
```

NAB. I think is map is more fitting here

Reply:
Agreed, changed.